src/wallet/wallet2.cpp

```diff
@@ -2856,18 +2845,51 @@ void wallet2::process_parsed_blocks(uint64_t start_height, const std::vector<cry
       THROW_WALLET_EXCEPTION_IF(txidx >= tx_cache_data.size(), error::wallet_internal_error, "txidx out of range");
       const cryptonote::transaction& tx = parsed_blocks[i].block.miner_tx;
       const size_t n_vouts = (m_refresh_type == RefreshType::RefreshOptimizeCoinbase && tx.version < 2) ? 1 : tx.vout.size();
-      tpool.submit(&waiter, [&, n_vouts, txidx](){ geniod(tx, n_vouts, txidx); }, true);
+      if (parsed_blocks[i].block.major_version >= HF_VERSION_VIEW_TAGS)
+        geniod_batch.push_back(geniod_params{ tx, n_vouts, txidx });
+      else
+        tpool.submit(&waiter, [&, n_vouts, txidx](){ geniod(tx, n_vouts, txidx); }, true);
```

Comment:
I saw a small (1-3%) perf hit scanning the whole chain when batching pre-view tags (reproduced on one of @gingeropolous' machines as well, thank you @gingeropolous).

I added `.reserve(num_txes)` to the vector to prevent it needing to resize -- it had no noticeable impact, but makes sense to add. See [this commit](https://github.com/monero-project/monero/pull/8061/commits/54231ebed3269873dd49ec229b86ff94a62e31c2).

_____

My hypothesis: because pre-view tag batches are so much slower to process, it ends up processing across fewer threads for longer. Imagine 1300 tx's in a set of blocks to process, 100 tx's per batch, 8 threads. The first 8 batches process roughly around the same time, then you have 5 batches remaining divided across 5 threads, which leaves 3 threads unused while the 5 remaining batches process slowly.

So, it would follow that submitting individual function calls to the thread pool seems to better allocate work across the available threads when the function being called is slow.

I tried exploring this theory further by calculating batch size relative to the number of threads available on the machine (instead of a fixed batch size) with no luck (this slowed it down, both on my machine and @Gingeropolous '). It feels worth exploring deeper at a future date I think. Here's what I tried for reference:

```
size_t TARGET_NUM_BATCHES_PER_THREAD = 100; // also tried 10
size_t max_concurrency = tpool.get_max_concurrency();
size_t num_batches = max_concurrency > 1 ? (TARGET_NUM_BATCHES_PER_THREAD * max_concurrency) : 1;
size_t geniods_per_batch = std::ceil(geniods.size() / static_cast<double>(num_batches));
```

Overall, I've consistently found that the approach in the PR as it is now empirically seems to be the best hybrid of fastest and simplest I could come up with, at least on the machines I've tested on.

Reply:
I wonder if there is overhead from first collecting all the geniods, then batching them. What if every time you collect batch_size txs, you dispatch them to the thread pool (you might need a std::move() to send them in, so you can make a new variable on the next loop)?

Reply:
Tried it [here](https://github.com/j-berman/monero/pull/4) and didn't observe a difference from first collecting then batching. It doesn't feel as clean code-wise to me, I could probably make it cleaner.